sirepo/template/cortex_sql_db.py

```diff
@@ -0,0 +1,86 @@
+"""Cortex db
+
+:copyright: Copyright (c) 2025 RadiaSoft LLC.  All Rights Reserved.
+:license: http://www.apache.org/licenses/LICENSE-2.0.html
+"""
+
+from pykern.pkcollections import PKDict
+from pykern.pkdebug import pkdc, pkdlog, pkdp
+import contextlib
+import pykern.sql_db
+import sqlalchemy
+import sirepo.sim_data
+import pykern.pkio
+
+_BASE = "cortex.sqlite3"
+
+
+def insert_material(parsed, qcall=None):
+    def _values(cols, vals):
+        return PKDict((c, vals[c]) for c in cols if c in vals)
+
+    with _session(qcall=qcall) as s:
+        i = s.insert(
+            "material",
+            # TODO(robnagler) need uid. we will have to have a db server of some sort
+            # it will have to validate incoming uid.
+            # sim_db_file is a model that could be used for writing to
+            # the database, because it validates the uid.
+            # sim_db_file should have a multithreaded worker queue. the serialization
+            # is already there.
+            _values(s.t.material.columns.keys(), parsed).pkupdate(uid="TODO RJN"),
+        ).material_id
+        for v in parsed.components.values():
+            s.insert(
+                "material_component",
+                _values(s.t.material_component.columns.keys(), v).pkupdate(
+                    material_id=i
+                ),
+            )
+
+
+@contextlib.contextmanager
+def _session(qcall):
+    s = sirepo.sim_data.get_class("cortex")
+    if w := (not (p := s.lib_file_exists(_BASE, qcall=qcall))):
+        p = pykern.pkio.py_path(_BASE)
+
+    with _meta(p).session() as rv:
+        yield rv
+    if w or s.hack_for_cortex_is_agent_side():
+        s.lib_file_write(_BASE, p, qcall=qcall)
+
+
+def _meta(path):
+    f = "float 64"
+    b = "bool"
+    u = "str 8"
+    return pykern.sql_db.Meta(
+        uri=pykern.sql_db.sqlite_uri(path),
+        schema=PKDict(
+            material=PKDict(
+                material_id="primary_id 1",
+                uid=u + " index",
+                created="datetime index",
+                material_name="str 100 unique",
+                availability_factor=f,
+                density_g_cm3=f,
+                is_atom_pct=b,
+                is_bare_tile=b,
+                is_homogenized_divertor=b,
+                is_homogenized_hcpb=b,
+                is_homogenized_wcll=b,
+                is_neutron_source_dt=b,
+                is_plasma_facing=b,
+                neutron_wall_loading="str 32",
+            ),
+            material_component=PKDict(
+                material_component_id="primary_id 2",
+                material_id="primary_id",
+                material_component_name="str 8 unique",
```

Comment:
material_component_name should be "unique" across the database, but only to that material_id.

If you upload a spreadsheet which shares components with another material, you get the error ```UNIQUE constraint failed: material_component.material_component_name```

Reply:
926956a